homeassistant/components/airzone/config_flow.py

```diff
@@ -76,3 +85,77 @@ async def async_step_user(
             data_schema=data_schema,
             errors=errors,
         )
+
+    async def async_step_dhcp(self, discovery_info: dhcp.DhcpServiceInfo) -> FlowResult:
+        """Handle DHCP discovery."""
+        self._discovered_ip = discovery_info.ip
+        self._discovered_mac = discovery_info.macaddress
+
+        _LOGGER.debug(
+            "DHCP discovery detected Airzone WebServer: %s", self._discovered_mac
+        )
+
+        self._async_abort_entries_match({CONF_HOST: self._discovered_ip})
+
+        await self.async_set_unique_id(format_mac(self._discovered_mac))
+        self._abort_if_unique_id_configured()
+
+        return await self.async_step_discovered_connection()
+
+    async def async_step_discovered_connection(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Confirm discovery."""
+        assert self._discovered_ip is not None
+        assert self._discovered_mac is not None
+
+        errors = {}
+        data_schema = vol.Schema(
+            {
+                vol.Required(CONF_PORT, default=DEFAULT_PORT): int,
+            }
+        )
+
+        if user_input is not None:
+            airzone = AirzoneLocalApi(
+                aiohttp_client.async_get_clientsession(self.hass),
+                ConnectionOptions(
+                    self._discovered_ip,
+                    user_input[CONF_PORT],
+                    user_input.get(CONF_ID, DEFAULT_SYSTEM_ID),
+                ),
+            )
+
+            try:
+                mac = await airzone.validate()
+            except InvalidSystem:
+                data_schema = data_schema.extend(
+                    {
+                        vol.Required(CONF_ID, default=1): int,
+                    }
+                )
```

Comment:
```suggestion
                data_schema[vol.Required(CONF_ID, default=1)] = int
```

This should work as well

Reply:
I'm afraid it doesn't work :$
```
            try:
                mac = await airzone.validate()
            except InvalidSystem:
>               data_schema[vol.Required(CONF_ID, default=1)] = int
E               TypeError: 'Schema' object does not support item assignment

homeassistant/components/airzone/config_flow.py:132: TypeError
```

Reply:
It likely has to be done before it's wrapped in vol.schema for it to work.  

I'll take a look later today and see if I can come up with a better suggestion.

Reply:
```diff
diff --git a/homeassistant/components/airzone/config_flow.py b/homeassistant/components/airzone/config_flow.py
index 3381b38da7..9fab2463fb 100644
--- a/homeassistant/components/airzone/config_flow.py
+++ b/homeassistant/components/airzone/config_flow.py
@@ -110,11 +110,7 @@ class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
         assert self._discovered_mac is not None
 
         errors = {}
-        data_schema = vol.Schema(
-            {
-                vol.Required(CONF_PORT, default=DEFAULT_PORT): int,
-            }
-        )
+        base_schema = {vol.Required(CONF_PORT, default=DEFAULT_PORT): int}
 
         if user_input is not None:
             airzone = AirzoneLocalApi(
@@ -129,11 +125,7 @@ class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
             try:
                 mac = await airzone.validate()
             except InvalidSystem:
-                data_schema = data_schema.extend(
-                    {
-                        vol.Required(CONF_ID, default=1): int,
-                    }
-                )
+                base_schema[vol.Required(CONF_ID, default=1)] = int
                 errors[CONF_ID] = "invalid_system_id"
             except AirzoneError:
                 errors["base"] = "cannot_connect"
@@ -156,6 +148,6 @@ class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
 
         return self.async_show_form(
             step_id="discovered_connection",
-            data_schema=data_schema,
+            data_schema=vol.Schema(base_schema),
             errors=errors,
         )
```

This should work

Reply:
Added in 4557c161d9d95708984eac242e0025a8fa009676